core/src/main/java/org/elasticsearch/common/settings/Settings.java

```diff
@@ -1032,7 +1032,110 @@ public Builder normalizePrefix(String prefix) {
          * set on this builder.
          */
         public Settings build() {
-            return new Settings(Collections.unmodifiableMap(map));
+            return new Settings(map);
+        }
+    }
+
+    // TODO We could use an FST internally to make things even faster and more compact
+    private static final class FilteredMap extends AbstractMap<String, String> { // pkg private for testing
+        private final Map<String, String> delegate;
+        private final Predicate<String> filter;
+        private final String prefix;
+        private int size = -1; // we cache that size since we have to iterate the entire set
+        @Override
+        public Set<Entry<String, String>> entrySet() {
+            Set<Entry<String, String>> delegateSet = delegate.entrySet();
+            AbstractSet<Entry<String, String>> filterSet = new AbstractSet<Entry<String, String>>() {
+
+                @Override
+                public Iterator<Entry<String, String>> iterator() {
+                    Iterator<Entry<String, String>> iter = delegateSet.iterator();
+
+                    return new Iterator<Entry<String, String>>() {
+                        private int numIterated;
+                        Entry<String, String> current;
+                        @Override
+                        public boolean hasNext() {
+                            if (numIterated == size) { // early terminate
+                                assert size != -1 : "size was never set: " + numIterated + " vs. " + size;
+                                return false;
+                            }
+                            while(iter.hasNext()) {
+                                if (filter.test((current = iter.next()).getKey())) {
+                                    numIterated++;
+                                    return true;
+                                }
+                            }
+                            current = null;
+                            return false;
+                        }
+
+                        @Override
+                        public Entry<String, String> next() {
```

Comment:
I don't think we can assume here that hasNext was called first. Also, it should throw a NoSuchElementException if there are no more entries as per the iterator.next API contract?